### DISCUSSION TITLE
How to use new "flat config" approach in Eslint?

### DISCUSSION THREAD

**Discussion by ryancwalsh** [vercel/next.js#49337] [Help] [answered]
### Summary

I have Eslint working in a TypeScript Next.js project, but I want to switch to the new "[flat config](https://eslint.org/blog/2022/08/new-config-system-part-2/)" approach that Eslint offers.

(Why is that not Next.js's default?)

I created `/eslint.config.js` to look something like this: https://stackoverflow.com/a/74819187/470749

Part 1
------

How can I migrate the `extends: ['next/core-web-vitals'],` to the flat config, which doesn't support `extends`?

Part 2
------

Other than missing `'next/core-web-vitals'`, if I run `eslint pages/*`, I see the errors that I expect.

But my VSCode no longer shows errors.

I see `"eslint.experimental.useFlatConfig": true,` in my VSC settings, and I've restarted VSC, but it still doesn't work.

How can I get the flat config to run automatically in VSC?

**Comment by milos-sikic-nimbus-tech** [comment]
This worked for me. I haven't tested it thoroughly, but it reports that I shouldn't use the `<img>` component, for example. If more people could test it out and provide feedback, I might be able to tweak it further.

```js
import nextPlugin from '@next/eslint-plugin-next';
import reactPlugin from 'eslint-plugin-react';
import hooksPlugin from 'eslint-plugin-react-hooks';

export default [
  {
    plugins: {
      react: reactPlugin,
    },
    rules: {
      ...reactPlugin.configs['jsx-runtime'].rules,
    },
    settings: {
      react: {
        version: 'detect', // You can add this if you get a warning about the React version when you lint
      },
    },
  },
  {
    plugins: {
      'react-hooks': hooksPlugin,
    },
    rules: hooksPlugin.configs.recommended.rules,
  },
  {
    plugins: {
      '@next/next': nextPlugin,
    },
    rules: {
      ...nextPlugin.configs.recommended.rules,
      ...nextPlugin.configs['core-web-vitals'].rules,
    },
  },
  {
    ignores: ['.next/*'],
  },
];
```

You need to install `@next/eslint-plugin-next`, `eslint-plugin-react` and `eslint-plugin-react-hooks` as dev dependencies.

**Reply by ryancwalsh** [reply]
Thanks for your response @milos-sikic-nimbus-tech .

I ran `yarn add -D @next/eslint-plugin-next eslint-plugin-react eslint-plugin-react-hooks`.

My only eslint settings in my VSC's `settings.json` are:

```JSON
"eslint.experimental.useFlatConfig": true,
  "eslint.debug": true,
  "eslint.options": {
    "overrideConfigFile": "eslint.config.mjs"
  },
  "settingsSync.ignoredSettings": [
    "eslint.experimental.useFlatConfig"
  ],
```

I've tried putting the config at `eslint.config.mjs` and `eslint.config.mjs`. I've tried with and without `"type": "module",` in my `package.json`.

`npx eslint` and `./node_modules/.bin/eslint` both give no output.

I have no `.eslintignore` file.

My guess is that you're not using TypeScript. Using your approach, I was able to get VSC to highlight errors in regular JS, but I never want to use JS. I always use TS and TSX now.

Here is my `eslint.config.js` currently:

```JavaScript
import nextPlugin from '@next/eslint-plugin-next';
import reactPlugin from 'eslint-plugin-react';
import hooksPlugin from 'eslint-plugin-react-hooks';
import ts from '@typescript-eslint/eslint-plugin';
import tsParser from '@typescript-eslint/parser';
import functional from 'eslint-plugin-functional';
import imprt from 'eslint-plugin-import'; // 'import' is ambiguous & prettier has trouble

const myRules = {
  // https://eslint.org/docs/latest/rules/max-len
  // Temporary rule to see if eslint will work:
  'max-len': ['warn', { code: 3, ignoreComments: false, ignoreRegExpLiterals: true, ignoreStrings: true, ignoreTemplateLiterals: true }],
};

export default [
  {
    rules: myRules,
  },
  {
    plugins: {
      react: reactPlugin,
    },
    rules: {
      ...reactPlugin.configs['jsx-runtime'].rules,
    },
    settings: {
      react: {
        version: 'detect', // You can add this if you get a warning about the React version when you lint
      },
    },
  },
  {
    plugins: {
      'react-hooks': hooksPlugin,
    },
    rules: hooksPlugin.configs.recommended.rules,
  },
  {
    plugins: {
      '@next/next': nextPlugin,
    },
    rules: {
      ...nextPlugin.configs.recommended.rules,
      ...nextPlugin.configs['core-web-vitals'].rules,
    },
  },
  {
    ignores: ['.next/*'],
  },
  {
    files: ['**/*.ts', '**/*.tsx'],
    languageOptions: {
      parser: tsParser,
      parserOptions: {
        ecmaFeatures: { modules: true },
        ecmaVersion: 'latest',
        project: './tsconfig.json',
      },
    },
    plugins: {
      functional,
      import: imprt,
      '@typescript-eslint': ts,
      ts,
    },
    rules: {
      ...ts.configs['eslint-recommended'].rules,
      ...ts.configs['recommended'].rules,
      ...myRules,
      'ts/return-await': 2,
    },
  },
];
```

`./node_modules/.bin/eslint .` shows warnings as I'd expect.

However, my VSC is not showing them as I type (the way it did before switching to flat config).

Thanks for any ideas.

**Reply by milos-sikic-nimbus-tech** [reply, marked as answer]
Hmm, it worked for me for TS and TSX files. Might be that it's missing the matcher?

We could maybe put the React and Next stuff into one object and add `files: ['**/*.ts', '**/*.tsx'],` to it.

Something like this:

```
export default [
  {
    files: ['**/*.ts', '**/*.tsx'],
    plugins: {
      react: reactPlugin,
      'react-hooks': hooksPlugin,
      '@next/next': nextPlugin,
    },
    rules: {
      ...reactPlugin.configs['jsx-runtime'].rules,
      ...hooksPlugin.configs.recommended.rules,
      ...nextPlugin.configs.recommended.rules,
      ...nextPlugin.configs['core-web-vitals'].rules,
      '@next/next/no-img-element': 'error',
    },
  },
  ... rest of your config
]
```

And yes, sorry, I didn't include one other plugin that we're using that's internal that might be helping. But I just took a look, and it's very similar to what you're doing, even the order in the array is the same.

Also, if you're not already, always restart the ESLint Server within VS Code when you change the config if you want to see the changes - it doesn't pick up on them all the time. Doing a restart guarantees that the newest config is loaded. Also, check out the output of the ESLint plugin (Cmd + Shift + P and type "ESLint output", you'll find the command).

Let me know if that helps!

**Reply by ryancwalsh** [reply]
Thank you so much, @milos-sikic-nimbus-tech !! 

I had already seen in the terminal `TypeError: Could not find the rule "canonical/filename-match-exported".`, but I'd ignored it.

Seeing it again in ESLint Show Output Channel made me wonder whether VSC was aborting there and not bothering to run any more rules.

And that seems to be the case.

Once I removed `/* eslint-disable canonical/filename-match-exported */` from the files that had those comments, VSC started honoring my ESLint rules (even in tsx files). 

Thanks!

**Reply by milos-sikic-nimbus-tech** [reply]
For the `canonical` rule, you're probably missing the plugin: `eslint-plugin-canonical`. 

Install that as a dev dependency

Import it as `import canonical from eslint-plugin-canonical`

And then just include it as a plugin

```
...
plugins: {
   canonical,
   ...
},
...
```

Do note that the name of the plugin is important here. In order for rules to be picked up, you need to specify a name for the plugin that's the same as the one that's used in the rules string, and comes before the `/` sign. This would also work:

```
import it as `import canonicalPlugin from eslint-plugin-canonical`

...
plugins: {
   canonical: canonicalPlugin
},
...
```

And according to the documentation, also this (haven't tested it):

```
import it as `import canonicalPlugin from eslint-plugin-canonical`

...
plugins: {
  canonicalPlugin
},
rules: {
   "canonicalPlugin/filename-match-exported": warn
}
...
```

Also, you're welcome! Glad it's working for you! :)

**Reply by ryancwalsh** [reply]
Thanks. Yeah, I partially got 'canonical' working like this: https://github.com/eslint/eslint/discussions/17228 But as I describe there, there is much more I'd love to get working. 

I might switch from `FlatCompat` to your approach too.

**Reply by nyxb** [reply]
hey maybe you can help me too. I can activate the rules only if I enter them individually the standard rules have no effect.

```ts
import type { ConfigItem, OptionsHasTypeScript, OptionsOverrides, OptionsStylistic } from '../types'

import { GLOB_NEXT } from '../globs'
import { parserTs, pluginNext, pluginReact, pluginReactHooks } from '../plugins'

export function next(
   options: OptionsHasTypeScript & OptionsOverrides & OptionsStylistic = {},
): ConfigItem[] {
   const {
      overrides = {},
      stylistic = true,
   } = options

   return [
      {
         name: 'nyxb:next:setup',
         plugins: {
            'next': pluginNext,
            'react': pluginReact,
            'react-hooks': pluginReactHooks,
         },
      },
      {
         files: [GLOB_NEXT],
         languageOptions: {
            parser: options.typescript ? parserTs as any : null,
            parserOptions: {
               ecmaFeatures: {
                  jsx: true,
               },
               sourceType: 'module',
            },
         },
         name: 'nyxb:next:rules',
         rules: {
            ...pluginReact.configs['jsx-runtime'].rules as any,
            ...pluginReactHooks.configs.recommended.rules as any,
            ...pluginNext.configs.recommended.rules as any,
            ...pluginNext.configs['core-web-vitals'].rules as any,
            'next/google-font-display': 'error',
            ...stylistic
               ? {
                  }
               : {},

            ...overrides,
         },
      },
   ]
}
```

in the screenshots you can see that only conditional rules are activated and in the next rules actually only the one that I have activated individually.

React: 
<img width="1401" alt="image" src="https://github.com/vercel/next.js/assets/98586665/3f9aa0fa-c73e-41bd-b5a7-2b2f7e099ecf">

React Hooks:
<img width="1407" alt="image" src="https://github.com/vercel/next.js/assets/98586665/63d0a740-cdd9-4de0-928e-402f9864f51b">

Next: 
<img width="1405" alt="image" src="https://github.com/vercel/next.js/assets/98586665/425a66ac-8c05-4a95-8b8d-414427b5c6f6">

all other are off:
<img width="1417" alt="image" src="https://github.com/vercel/next.js/assets/98586665/158bd222-bc87-48be-a2c6-f176109b8c0a">


the packages are all on the latest version

**Reply by mirismaili** [reply]
@milos-sikic-nimbus-tech 
And [the rules of eslint-config-next](https://github.com/vercel/next.js/blob/d4fcd039666caefe2a69bd22f487f6b694d595c3/packages/eslint-config-next/index.js#L62-L78)?

**Reply by rohangupta-zomato** [reply]
Hey @nyxb Which tool are you using this which is listing all rules like that?

**Reply by nyxb** [reply]
> Hey @nyxb Which tool are you using this which is listing all rules like that?

https://github.com/eslint/config-inspector

**Comment by bestickley** [comment]
@ryancwalsh, did you consider using `FlatCompat` from @eslint/eslintrc like below?
```ts
import path from "node:path";
import { fileURLToPath } from "url";
import { FlatCompat } from "@eslint/eslintrc";
import { defineFlatConfig } from "eslint-define-config";
import ts from "@typescript-eslint/eslint-plugin";
import tsParser from "@typescript-eslint/parser";
import js from "@eslint/js";

const baseDirectory = path.dirname(fileURLToPath(import.meta.url));

const compat = new FlatCompat({
  baseDirectory,
  resolvePluginsRelativeTo: baseDirectory,
});

export default defineFlatConfig([
{
    files: ["**/*.{ts,tsx}"],
    plugins: {
      "@typescript-eslint": ts,
    },
    languageOptions: {
      parser: tsParser,
    },
    rules: {
      ...js.configs.recommended.rules,
      ...ts.configs.recommended.rules,
    },
  },
  compat.config({
      overrides: [
        { files: "ui/**/*.{ts,tsx}", extends: "next/core-web-vitals" },
      ],
    })
]);
```
I'm having issues getting this to work, but theoretically I think this is the ideal solution.

**Reply by ryancwalsh** [reply]
I see from https://github.com/vercel/next.js/discussions/49337#discussioncomment-6030437 that I was using FlatCompat, but it has been a long time, so I don't remember anything that isn't documented above.

If I recall, I probably kept using original eslint configs instead of the new flat style. Switching my projects to flat config was taking more time than it was worth.

**Comment by jeremy-code** [comment]
EDIT again:

I have created a package [`eslint-config-next-flat`](https://www.npmjs.com/package/eslint-config-next-flat) ([Source code](https://github.com/jeremy-code/eslint-config-next-flat)) that is (roughly) a port of the eslint-config-next for FlatConfig. It seems to work perfect for TypeScript projects (not sure about native JavaScript since it requires Babel parser and I don't have a complicated vanilla JS Next.js project to try it out with). 

Just another option whilst we wait for the updated config.

---

EDIT:

After some more research, I believe the best options thus far until Next.js releases a flat config is:

eslint.config.mjs
```js

// flat compat
import url from "node:url";
import eslint from "@eslint/js";
import { FlatCompat } from "@eslint/eslintrc";

const __dirname = url.fileURLToPath(new URL(".", import.meta.url));
const compat = new FlatCompat({ baseDirectory: __dirname });

/** @type {import('eslint').Linter.FlatConfig[]} */
export default [eslint.configs.recommended, ...compat.extends("next")];
```

or (but you miss out on setting other eslint rules and parsing, and just get access to the default nextjs rules)
```js
import eslint from "@eslint/js";
import next from "@next/eslint-plugin-next";

/** @type {import('eslint').Linter.FlatConfig[]} */
export default [
  eslint.configs.recommended,
  {
    plugins: {
      "@next/next": next,
    },
    rules: {
      ...next.configs.recommended.rules,
    },
  },
];

```

you can also just use the legacy config and not upgrade to eslint v9, which probably is the best idea for maximum compatability (i had some issues using FlatCompat in a monorepo)

---

Hello, sorry for reviving an answered discussion but I have discovered an (external) package that offers the Next.js rules for use in ESLint Flat config. Not affiliated with them at all -- discovered it in use on the discord.js repo ([discord.js/eslint.config.js](https://github.com/discordjs/discord.js/blob/main/eslint.config.js#L4)).

`import next from 'eslint-config-neon/flat/next.js';`

from ESLint Config Neon: https://github.com/iCrawl/eslint-config-neon

FlatCompat is also an option

**Comment by mrcaidev** [comment]
ESLint has just released [@eslint/compat](https://eslint.org/blog/2024/05/eslint-compatibility-utilities/), which looks much more elegant.

```js
// eslint.config.js
import { fixupConfigRules } from "@eslint/compat";
import { FlatCompat } from "@eslint/eslintrc";

const compat = new FlatCompat();

/** @type {import("eslint").Linter.FlatConfig[]} */
const config = [
  ...fixupConfigRules(compat.extends("next/core-web-vitals", "prettier")),
  { ignores: [".next"] },
];

export default config;
```

This works fine, until `--cache` is passed to ESLint, when ESLint would fail and give the following error. Any ideas?

```
Error: Could not serialize parser object (missing 'meta' object).
    at Object.value (/home/mrcaidev/oauth/node_modules/.pnpm/eslint@9.3.0/node_modules/eslint/lib/config/flat-config-array.js:325:27)
    at stringify (/home/mrcaidev/oauth/node_modules/.pnpm/json-stable-stringify-without-jsonify@1.0.1/node_modules/json-stable-stringify-without-jsonify/index.js:25:25)
    at module.exports (/home/mrcaidev/oauth/node_modules/.pnpm/json-stable-stringify-without-jsonify@1.0.1/node_modules/json-stable-stringify-without-jsonify/index.js:68:7)
    at hashOfConfigFor (/home/mrcaidev/oauth/node_modules/.pnpm/eslint@9.3.0/node_modules/eslint/lib/cli-engine/lint-result-cache.js:50:75)
    at LintResultCache.getCachedLintResults (/home/mrcaidev/oauth/node_modules/.pnpm/eslint@9.3.0/node_modules/eslint/lib/cli-engine/lint-result-cache.js:116:30)
    at /home/mrcaidev/oauth/node_modules/.pnpm/eslint@9.3.0/node_modules/eslint/lib/eslint/eslint.js:896:41
    at Array.map (<anonymous>)
    at ESLint.lintFiles (/home/mrcaidev/oauth/node_modules/.pnpm/eslint@9.3.0/node_modules/eslint/lib/eslint/eslint.js:868:23)
    at async Object.execute (/home/mrcaidev/oauth/node_modules/.pnpm/eslint@9.3.0/node_modules/eslint/lib/cli.js:500:23)
    at async main (/home/mrcaidev/oauth/node_modules/.pnpm/eslint@9.3.0/node_modules/eslint/bin/eslint.js:153:22)
```

**Reply by Alarakov** [reply]
guys how to use this flat config with next lint command? when i write next lint it asked to create a config?

**Reply by r34son** [reply]
Having same issue

**Reply by dir** [reply]
I just implemented the fixConfigRules + compat setup that you're using (thank you for the snippet by the way). I can't seem to replicate the `--cache`  issue that you're having. The setup below is working quite well for me. I am using `npx eslint . --cache`.

```js
// @ts-check

import path from 'path';
import { fileURLToPath } from 'url';

import { fixupConfigRules } from '@eslint/compat';
import { FlatCompat } from '@eslint/eslintrc';
import eslint from '@eslint/js';
import prettierConfig from 'eslint-config-prettier';
import tseslint from 'typescript-eslint';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const compat = new FlatCompat({
  baseDirectory: __dirname,
});

export default tseslint.config(
  {
    ignores: [
      '.next/**',
      'node_modules/**',
      'tailwind.config.ts',
      'src/env.mjs',
      'next.config.js',
      'next.config.mjs',
      'prettier.config.cjs',
      'postcss.config.cjs',
    ],
  },
  eslint.configs.recommended,
  ...fixupConfigRules(compat.extends('next/core-web-vitals')),
  ...tseslint.configs.recommendedTypeChecked,
  ...tseslint.configs.stylisticTypeChecked,
  prettierConfig,
  {
    plugins: {
      '@typescript-eslint': tseslint.plugin,
    },
    languageOptions: {
      parser: tseslint.parser,
      parserOptions: {
        project: './tsconfig.json',
        tsconfigRootDir: import.meta.dirname,
      },
    },
  },
);

```

**Reply by jkaleshi14** [reply]
when trying @dir step's above I get the following error:

`
ESLint: 9.5.0

Error: Cannot read config file: /node_modules/.pnpm/eslint-config-next@14.2.3_eslint@9.5.0_typescript@5.4.5/node_modules/eslint-config-next/index.js
Error: The ESLint patch script has only been tested with ESLint version 6.x, 7.x, and 8.x. (Your version: 9.5.0)
Consider reporting a GitHub issue:
https://github.com/microsoft/rushstack/issues
Referenced from: /node_modules/.pnpm/eslint-config-next@14.2.3_eslint@9.5.0_typescript@5.4.5/node_modules/eslint-config-next/core-web-vitals.js
    at Object.<anonymous> (/node_modules/.pnpm/@rushstack+eslint-patch@1.10.2/node_modules/@rushstack/eslint-patch/lib/_patch-base.js:184:11)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/node_modules/.pnpm/@rushstack+eslint-patch@1.10.2/node_modules/@rushstack/eslint-patch/lib/modern-module-resolution.js:11:23)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
`

**Reply by dir** [reply]
A couple differences between my setup and yours I notice off the bat @jkaleshi14:

1. I am using eslint 8.57.0
2. I'm using NPM, not pnpm
3. I can't quite tell, but it looks like you're using CJS. My eslint config is `eslint.config.mjs`.

I have a hunch that addressing 3 on your end might fix it. But, could be one of the many other externalities as well.

**Reply by jkaleshi14** [reply]
@dir Thanks that fixed my issue

**Reply by r34son** [reply]
> @dir Thanks that fixed my issue

Could you share fix, please

**Reply by dir** [reply]
@r34son I believe I already did, refer to my previous message.

**Reply by timtucker-dte** [reply]
See the notes here on the root cause for why the errors about meta are occurring -- Next.js isn't following the spec for custom parsers and is omitting the "meta" section when it re-exports the parser from @babel/eslint-parser:
https://github.com/eslint/eslint/issues/19080

Tried to report as an issue, but it got auto-closed because the Next.js repository apparently isn't considered a valid link to reproduce.
https://github.com/vercel/next.js/issues/72833

**Comment by dir** [comment]
Wish the Vercel team prioritized this a bit higher. It's a fairly small lift but is impacting the developer experience for a lot of us. 

If Next is striving so aggressively to use the newest standards (RSC, server actions, etc), can we not reflect that same mindset with things as small as our ESLint configs?

**Comment by poksme** [comment]
I wrote a setup guide for Eslint 9 with Next.js detailing my approach to flat config with Next.js : [Eslint 9 & Next.js — Setup Guide](https://medium.com/linotte-technology/eslint-9-next-js-935c2b6d0371)

Also for a TLDR; check my comment [here](https://github.com/vercel/next.js/issues/64409#issuecomment-2265632495)

**Comment by boning-w** [comment]
For anyone who want try eslint 9 flat config in Nextjs. Make sure to backup your old `.eslintrc.json` file, then try:

See: https://eslint.org/docs/latest/use/configure/migration-guide#migrate-your-config-file

```bash
npx @eslint/migrate-config .eslintrc.json
```


This will create a `eslint.config.mjs` for you. 

Then install some packages `@eslint/eslintrc`, `@eslint/js`, etc.

Then run `npm run lint` to see if it work.

Works for me in Next.js 15.

```json
  "dependencies": {
    "next": "^15.0.1",
    "react": "^18",
    "react-dom": "^18"
  },
  "devDependencies": {
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "@eslint/eslintrc": "^3.1.0",
    "@eslint/js": "^9.13.0",
    "eslint": "^9.13.0",
    "eslint-config-next": "^15.0.1",
    "postcss": "^8",
    "tailwindcss": "^3.4.1"
  }
```

**Reply by mdillgen** [reply]
@boning-w The migration went smoothly for me. Even the paths of the deprecated .eslintignore are taken into account!

**Reply by bdlowery** [reply]
Wow. This is SO MUCH BETTER than the .eslintrc.json version.

Output from `npx @eslint/migrate-config .eslintrc.json`
```js
import path from "node:path";
import { fileURLToPath } from "node:url";
import js from "@eslint/js";
import { FlatCompat } from "@eslint/eslintrc";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const compat = new FlatCompat({
    baseDirectory: __dirname,
    recommendedConfig: js.configs.recommended,
    allConfig: js.configs.all
});

export default [...compat.extends("next/core-web-vitals"), {
    rules: {
        "react/no-unescaped-entities": "warn",
    },
}];
```
versus my .eslintrc.json file:
```js
{
	"extends": "next/core-web-vitals",
	"rules": {
		"react/no-unescaped-entities": "warn"
	}
}
```

Not. This sucks. I'll just do `ESLINT_USE_FLAT_CONFIG=false.`

**Reply by isaacbowen** [reply]
😂 😂 😂

**Comment by corradin** [comment]
Soooo, on 18 March 2025 a [named export](https://github.com/acdlite/next.js/blob/7c6f6049b631439482f77c5d51b78022f559afac/packages/eslint-plugin-next/src/index.ts#L70) has been added, allowing you to do this:

```js
import { flatConfig } from '@next/eslint-plugin-next';

export default {
  ...flatConfig.recommended,
  ...flatConfig.coreWebVitals,
};
```

Thank you @SukkaW and @gaojude 😍😍😍

**Reply by ya2s** [reply]
How will [next/typescript](https://nextjs.org/docs/app/api-reference/config/eslint#with-typescript) be handled?
Will the document not be updated to this style?

**Reply by SukkaW** [reply]
> next/typescript

@ya2s `next/typescript` lives in `eslint-config-next`, not in `eslint-plugin-next`. Notice the difference between `-config-` and `-plugin-` here.

`eslint-config-next` has some extra hacks that can not be migrated to flat configs.

**Comment by ShueiYang** [comment]
Sooo in case people try to set up the flat config with the new `typescript-eslint` instead of `next/typescript` at this moment  looks like this setup work:

```ts
import eslint from "@eslint/js";
import tseslint from "typescript-eslint";
import { flatConfig } from "@next/eslint-plugin-next";
import eslintPluginReactHooks from "eslint-plugin-react-hooks";
import eslintPluginPrettierRecommended from "eslint-plugin-prettier/recommended";

export default tseslint.config(
  flatConfig.recommended,
  flatConfig.coreWebVitals,
  {
    files: ["**/*.ts", "**/*.tsx"],

    extends: [
      eslint.configs.recommended,
      tseslint.configs.recommended, // or more strict config (e.g. `configs.recommendedTypeChecked`)
      //... other configs 
      // For my useCase I put on the last `eslint-plugin-prettier/recommanded` that combine in one go 
      // with `eslint-config-prettier` So i didn't install eslint-config-prettier.
      eslintPluginPrettierRecommended,
    ],
    
   // Skip this if you don't use `eslint-plugin-react-hooks`
    plugins: {
      "react-hooks": eslintPluginReactHooks,
    },

    languageOptions: {
      // TypeScript ESLint parser for TS files
      parser: tseslint.parser,
      parserOptions: {
        project: "./tsconfig.json",
        ecmaVersion: "latest",
        sourceType: "module",
        ecmaFeatures: {
          jsx: true,
        },
        // the 4 keys above in the parserOptions may be replaced with the 2 options below
        // projectService: true,
        // tsconfigRootDir: import.meta.dirname,
      },
    },
   
   // Custom rule to overwrite/modify or to disable if necessary 
    rules: {
      //...
    },
  },
);
```

**Reply by quanghuynt14** [reply]
I was exhausted from dealing with ESLint setup when I came across your comment. I did the exact same thing but ran into this error:
`ConfigError: Config (unnamed): Key "extends": This appears to be in eslintrc format rather than flat config format.`

Here's how I fixed it:
```js
import eslintPluginNext from "@next/eslint-plugin-next";
import tseslint from "typescript-eslint";
import eslintPluginReactHooks from "eslint-plugin-react-hooks";
import eslintPluginPrettierRecommended from "eslint-plugin-prettier/recommended";

const eslintConfig = [
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  eslintPluginNext.flatConfig.recommended,
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  eslintPluginNext.flatConfig.coreWebVitals,
  ...tseslint.configs.recommendedTypeChecked,
  eslintPluginPrettierRecommended,
  {
    plugins: {
      "react-hooks": eslintPluginReactHooks,
    },

    languageOptions: {
      parser: tseslint.parser,
      parserOptions: {
        project: "./tsconfig.json",
        ecmaVersion: "latest",
        sourceType: "module",
        ecmaFeatures: {
          jsx: true,
        },
      },
    },
  },
];

export default eslintConfig;
```

**Reply by ShueiYang** [reply]
Oh ok 🤔 you have this error with the `config` method ? in this case maybe you can also do something like this:

```ts
 export default tseslint.config(
   eslintPluginNext.flatConfig.recommended,
   eslintPluginNext.flatConfig.coreWebVitals,
   tseslint.configs.recommendedTypeChecked,  // or spread...

  //...
```
This is how i set up without framework with React and Vite, the `config`  method is optional ofc, they explain why they recommend it from there doc here https://typescript-eslint.io/packages/typescript-eslint#config
